// stores/StoreWrapper.ts
import { createStore, Store, AnyAction, Unsubscribe, Reducer } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';

export default class StoreWrapper<State, Actions, WState = null> {
    protected store: Store<State>;
    public actions: Actions;

    constructor(
        protected reducer: Reducer,
        protected actionMap: {[key: string]: (...any) => void},
        protected name: string,
        protected stateWrapper = null as (state: State) => State & WState
    ) {
        let middleWere = undefined;

        if (process.env.NODE_ENV === 'development') {
            middleWere = composeWithDevTools({ name: `${ window.location.host }:${ this.name }` })();
        }

        /******************************/

        this.store = createStore(
            this.reducer,
            middleWere
        );

        /******************************/

        let actions = {};

        for (let key in actionMap) {
            actions[ key ] = actionMap[ key ].bind(this.store);
        }

        this.actions = actions as Actions;
    }

    /******************************/

    getState(): State {
        return this.stateWrapper !== null
            ? this.stateWrapper(this.store.getState())
            : this.store.getState();
    }
    getStore(): Store<State> {
        return this.store;
    }
    dispatch(action: AnyAction): AnyAction {
        return this.store.dispatch(action);
    }
    subscribe(listener: () => void): Unsubscribe {
        return this.store.subscribe(listener);
    }
}
