// stores/global/types.ts
export type State = {
    settings: {
        dontSpy: boolean,
        gdrpShown: boolean,
        map: {
            colorBackground: string,
            colorBorderProvince: string,
            colorBorderContinent: string,
            colorVillagePlayer: string,
            colorVillageBarbarian: string,
            colorHoverItem: string,
            colorHoverGroup: string
        }
    }
};

export const INIT_STATE: State = {
    settings: {
        dontSpy: false,
        gdrpShown: false,
        map: {
            colorBackground: '#004000',
            colorBorderProvince: '#888888',
            colorBorderContinent: '#ffff00',
            colorVillagePlayer: '#003000',
            colorVillageBarbarian: '#005000',
            colorHoverItem: '#ffffff',
            colorHoverGroup: '#999999'
        }
    }
};

/******************************/

export const IMPORT_DB = 'global/import/db';
export const SET_DONT_SPY = 'settings/dontSpy';
export const SET_GDRP_SHOWN = 'settings/gdrpShown';
