// view/pages/List.tsx
import * as React from 'react';
import { RouteComponentProps, match } from 'react-router-dom';
import { ViewContext } from 'view/viewContext';
import Header from 'view/pages/list/Header';
import Footer from 'view/pages/list/Footer';
import Info from 'view/pages/list/Info';
import Market from 'view/pages/list/list/Marker';
import BtnBackToHome from 'view/components/BtnBackToHome';
import World from 'view/pages/list/list/World';
import { IconSettings } from 'view/components/icon';

/******************************/

const MARKET_NAME = {
    'beta': 'Beta',
    'en': 'United Kingdom',
    'us': 'USA',
    'de': 'Germany',
    'nl': 'Netherlands',
    'fr': 'France',
    'pl': 'Poland',
    'ru': 'Russia',
    'br': 'Brazil',
    'es': 'Spain',
    'it': 'Italy',
    'gr': 'Greece',
    'tr': 'Turkey',
    'cz': 'Czech Republic',
    'pt': 'Portugal',
    'se': 'Sweden',
    'no': 'Norway',
    'dk': 'Denmark',
    'fi': 'Finland',
    'sk': 'Slovakia',
    'ro': 'Romania',
    'hu': 'Hungary'
};

/******************************/

type MarketData = {
    market: string,
    worldList: any[]
};

type ComponentProps = RouteComponentProps & Readonly<{
    match: match<{ market: null }>,
    context: ViewContext,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{
    mapList: MarketData[]
}>;

/******************************/

export default
class PageList
extends React.Component<ComponentProps, ComponentState> {
    // callback do unmount - nasłuch adresu
    unlinstenHistory: Function;

    constructor(props) {
        super(props);

        this.state = {
            mapList: []
        }

        /******************************/

        this.showSettings = this.showSettings.bind(this);
    }

    /******************************/

    showSettings() {
        this.props.context.openSettings('analytics');
    }

    setMap(market: string, worldId: string): void {
        this.props.history.push(`/${ market }/${ worldId }`);
    }
    setMarket(market: string): void {
        this.props.history.push(`/${ market }`);
    }

    /******************************/

    componentDidMount() {
        const refWindow = window as typeof window & {initGoogleAdSense(): void};
        const { hostname } = location;

        const parseRequest = (res) => {
            if (res.status === 200) {
                res.json().then(
                    (jsonData) => this.setState({mapList: jsonData}),
                    (err) => null
                );
            }
        }

        fetch(`/data/data.json`).then(
            parseRequest,
            () => fetch(`/mapList`).then(parseRequest, () => {})
        );

        this.unlinstenHistory = this.props.history.listen((location, action) => {
            this.props.context.modalSystem.closeAll();
        });

        if ('initGoogleAdSense' in window) {
            refWindow.initGoogleAdSense();
        }
    }

    componentWillUnmount() {
        this.unlinstenHistory();
    }

    /******************************/

    render(): React.ReactNode {
        let mapList = [],
            { market } = this.props.match.params;

        if (market) {
            // lista światów
            for (let i = 0; i < this.state.mapList.length; i++) {
                let marketData: MarketData = this.state.mapList[ i ],
                    marketKey: string = marketData.market;

                if (market === marketKey) {
                    let worldList: any = marketData.worldList;

                    mapList.push(
                        <div
                            key={'header'}
                            className="d-flex justify-content-between mb-2"
                        >
                            <img className="img-flag" src={ `/flags/${ marketKey }.svg` } />
                            &nbsp;
                            { MARKET_NAME[ marketKey ] || marketKey }
                            <BtnBackToHome
                                history={ this.props.history }
                            />
                        </div>
                    );

                    for (let i = 0; i < worldList.length; i++) {
                        let worldData = worldList[ i ];

                        mapList.push(
                            <World
                                key={ worldData.id }
                                worldId={ worldData.id }
                                worldName={ worldData.name }
                                onClick={ () => this.setMap(market, worldData.id) }
                            />
                        );
                    }
                    break;
                }
            }
        } else {
            mapList.push(
                <div
                    key={'header'}
                    className="d-flex justify-content-between mb-2"
                >
                    Available markets
                </div>
            );

            // lista rynków
            for (let i = 0; i < this.state.mapList.length; i++) {
                let marketData: MarketData = this.state.mapList[ i ],
                    market: string = marketData.market;

                if (market === 'undefined') {
                    continue;
                }

                mapList.push(
                    <Market
                        key={ market }
                        marketId={ market }
                        onClick={
                            () => this.setMarket(market)
                        }
                    />
                );
            }
        }

        return (
            <div className="container position-relative">
                <div style={{
                    position: 'absolute',
                    top: 0,
                    left: 0,
                    width: '100%',
                    textAlign: 'right'
                }}>
                    <button
                        className="btn btn-primary"
                        onClick={ this.showSettings }
                        role="button"
                        aria-pressed="false"
                        aria-label="Settings"
                    >
                        <IconSettings />
                    </button>
                </div>
                <Header />
                <main className="row">
                    <div className="col-8">
                        <Info />
                    </div>
                    <div className="col-4">
                        <div className="jumbotron p-3">
                            { mapList }
                        </div>
                        <div className="jumbotron p-3">
                            <h5>
                                <img
                                    className="img-flag"
                                    src="/flags/tr.svg"
                                    alt="Turkey"
                                />&nbsp;Turkey
                            </h5>
                            <p>
                                Please unban the <kbd>TribalWars2Map</kbd> account.
                            </p>
                        </div>
                        <div className="jumbotron p-3">
                            <h4>Do you known?</h4>
                            <p>
                                Illegal scripts to Tribal Wars 2 can be blockked using two short lines of code?
                            </p>
                            <p>
                                To the <kbd>game_XXXXX.js</kbd> just add the code <kbd>(function()&#123;</kbd> to start and the <kbd>&#125;)();</kbd> to end the file.
                            </p>
                            <p>
                                But someone do nothing with that ☹️
                            </p>
                        </div>
                    </div>
                </main>
                <Footer />
            </div>
        );
    }
}
