// view/pages/List.tsx
import * as React from 'react';
import { RouteComponentProps, match } from 'react-router-dom';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import ModalManager from 'libs/ModalManager';
import featureSettingsInit from 'view/features/modalSettings';
import Header from 'view/pages/list/Header';
import Footer from 'view/pages/list/Footer';
import Info from 'view/pages/list/Info';
import Market from 'view/pages/list/list/Marker';
import BtnBackToHome from 'view/components/BtnBackToHome';
import World from 'view/pages/list/list/World';

/******************************/

const MARKET_NAME = {
    'beta': 'Beta',
    'en': 'United Kingdom',
    'us': 'USA',
    'de': 'Germany',
    'nl': 'Netherlands',
    'fr': 'France',
    'pl': 'Poland',
    'ru': 'Russia',
    'br': 'Brazil',
    'es': 'Spain',
    'it': 'Italy',
    'gr': 'Greece',
    'tr': 'Turkey',
    'cz': 'Czech Republic',
    'pt': 'Portugal',
    'se': 'Sweden',
    'no': 'Norway',
    'dk': 'Denmark',
    'fi': 'Finland',
    'sk': 'Slovakia',
    'ro': 'Romania',
    'hu': 'Hungary'
};

const mapListTest = [{"market":"beta","worldList":[{"id":"zz7","name":"Tempus","active":true,"lastUpdated":1568501576241},{"id":"zz8","name":"Semper","active":true,"lastUpdated":1568501576241}]},{"market":"de","worldList":[{"id":"de33","name":"Gaillard","active":false,"lastUpdated":1565234072004},{"id":"de34","name":"Histria","active":false,"lastUpdated":1565234072004},{"id":"de35","name":"Inveraray","active":false,"lastUpdated":1565234072004},{"id":"de36","name":"Jasenov","active":true,"lastUpdated":1567092751824},{"id":"de37","name":"Krak des Chevaliers","active":false,"lastUpdated":1565234072004},{"id":"de38","name":"Landskrona","active":false,"lastUpdated":1567092751824},{"id":"de39","name":"Mosteiro da Batalha","active":true,"lastUpdated":1567092751824},{"id":"de40","name":"Navardun","active":true,"lastUpdated":1567092751824},{"id":"de41","name":"Olavinlinna","active":true,"lastUpdated":1567092751824},{"id":"de42","name":"Palamidi","active":true,"lastUpdated":1567092751824},{"id":"de43","name":"Queen's Sconce","active":true,"lastUpdated":1567092751824}]},{"market":"ru","worldList":[{"id":"ru35","name":"Inveraray","active":false,"lastUpdated":1565234159761},{"id":"ru36","name":"Jasenov","active":false,"lastUpdated":1565234159761},{"id":"ru37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567092840024},{"id":"ru38","name":"Landskrona","active":true,"lastUpdated":1567092840024},{"id":"ru39","name":"Mosteiro da Batalha","active":true,"lastUpdated":1567092840024},{"id":"ru40","name":"Navardun","active":true,"lastUpdated":1567092840024},{"id":"ru41","name":"Olavinlinna","active":true,"lastUpdated":1567092840024}]},{"market":"es","worldList":[{"id":"es30","name":"Deva","active":false,"lastUpdated":1565234218409},{"id":"es31","name":"Eketorp","active":false,"lastUpdated":1565234218409},{"id":"es32","name":"Frankenstein","active":false,"lastUpdated":1567092884740},{"id":"es33","name":"Gaillard","active":true,"lastUpdated":1567092884740},{"id":"es34","name":"Histria","active":true,"lastUpdated":1567092884740},{"id":"es35","name":"Inveraray","active":true,"lastUpdated":1567092884740},{"id":"es36","name":"Jasenov","active":true,"lastUpdated":1567092884740},{"id":"es37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567092884740}]},{"market":"it","worldList":[{"id":"it25","name":"Yburg","active":false,"lastUpdated":1565234254606},{"id":"it28","name":"Bergenhus","active":false,"lastUpdated":1565234254606},{"id":"it29","name":"Castelo de Guimarães","active":false,"lastUpdated":1565234254606},{"id":"it30","name":"Deva","active":false,"lastUpdated":1565234254606},{"id":"it31","name":"Eketorp","active":false,"lastUpdated":1565234254606},{"id":"it32","name":"Frankenstein","active":false,"lastUpdated":1565234254606},{"id":"it33","name":"Gaillard","active":false,"lastUpdated":1567092913483},{"id":"it34","name":"Histria","active":true,"lastUpdated":1567092913483},{"id":"it35","name":"Inveraray","active":true,"lastUpdated":1567092913483},{"id":"it36","name":"Jasenov","active":true,"lastUpdated":1567092913483},{"id":"it37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567092913483}]},{"market":"en","worldList":[{"id":"en39","name":"Mosteiro da Batalha","active":false,"lastUpdated":1565234299153},{"id":"en41","name":"Olavinlinna","active":false,"lastUpdated":1565234299153},{"id":"en42","name":"Palamidi","active":false,"lastUpdated":1565234299153},{"id":"en43","name":"Queen's Sconce","active":true,"lastUpdated":1567092933531},{"id":"en44","name":"Rupea","active":true,"lastUpdated":1567092933531},{"id":"en45","name":"Stargard","active":true,"lastUpdated":1567092933531},{"id":"en46","name":"Tzschocha","active":true,"lastUpdated":1567092933531},{"id":"en47","name":"Uhrovec","active":true,"lastUpdated":1567092933531}]},{"market":"fr","worldList":[{"id":"fr33","name":"Gaillard","active":false,"lastUpdated":1565234385175},{"id":"fr34","name":"Histria","active":false,"lastUpdated":1565234385176},{"id":"fr35","name":"Inveraray","active":false,"lastUpdated":1565234385176},{"id":"fr36","name":"Jasenov","active":true,"lastUpdated":1567093014020},{"id":"fr37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567093014020},{"id":"fr38","name":"Landskrona","active":true,"lastUpdated":1567093014020},{"id":"fr39","name":"Mosteiro da Batalha","active":true,"lastUpdated":1567093014020},{"id":"fr40","name":"Navardun","active":true,"lastUpdated":1567093014020},{"id":"fr41","name":"Olavinlinna","active":true,"lastUpdated":1567093014020}]},{"market":"nl","worldList":[{"id":"nl29","name":"Castelo de Guimarães","active":false,"lastUpdated":1565230755800},{"id":"nl30","name":"Deva","active":true,"lastUpdated":1567093066946},{"id":"nl31","name":"Eketorp","active":true,"lastUpdated":1567093066946},{"id":"nl32","name":"Frankenstein","active":true,"lastUpdated":1567093066946},{"id":"nl33","name":"Gaillard","active":true,"lastUpdated":1567093066946},{"id":"nl34","name":"Histria","active":false,"lastUpdated":1567093066946},{"id":"nl35","name":"Inveraray","active":true,"lastUpdated":1567093066946},{"id":"nl36","name":"Jasenov","active":true,"lastUpdated":1567093066946},{"id":"nl37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567093066946}]},{"market":"pl","worldList":[{"id":"pl35","name":"Inveraray","active":false,"lastUpdated":1565230823282},{"id":"pl36","name":"Jasenov","active":false,"lastUpdated":1565230823282},{"id":"pl37","name":"Krak des Chevaliers","active":false,"lastUpdated":1565230823282},{"id":"pl38","name":"Landskrona","active":false,"lastUpdated":1567093118305},{"id":"pl39","name":"Mosteiro da Batalha","active":true,"lastUpdated":1567093118305},{"id":"pl40","name":"Navardun","active":true,"lastUpdated":1567093118305},{"id":"pl41","name":"Olavinlinna","active":true,"lastUpdated":1567093118305},{"id":"pl42","name":"Palamidi","active":true,"lastUpdated":1567093118305}]},{"market":"us","worldList":[{"id":"us32","name":"Frankenstein","active":false,"lastUpdated":1565230874676},{"id":"us35","name":"Inveraray","active":false,"lastUpdated":1565230874676},{"id":"us36","name":"Jasenov","active":false,"lastUpdated":1565230874676},{"id":"us37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567093166004},{"id":"us38","name":"Landskrona","active":false,"lastUpdated":1565230874676},{"id":"us39","name":"Mosteiro da Batalha","active":true,"lastUpdated":1567093166004},{"id":"us40","name":"Navardun","active":true,"lastUpdated":1567093166004},{"id":"us41","name":"Olavinlinna","active":true,"lastUpdated":1567093166004},{"id":"us42","name":"Palamidi","active":true,"lastUpdated":1567093166004}]},{"market":"gr","worldList":[{"id":"gr31","name":"Eketorp","active":false,"lastUpdated":1565230967462},{"id":"gr32","name":"Frankenstein","active":false,"lastUpdated":1565230967462},{"id":"gr33","name":"Gaillard","active":false,"lastUpdated":1565230967462},{"id":"gr34","name":"Histria","active":true,"lastUpdated":1567093280335},{"id":"gr35","name":"Inveraray","active":true,"lastUpdated":1567093280335},{"id":"gr36","name":"Jasenov","active":true,"lastUpdated":1567093280335},{"id":"gr37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567093280335}]},{"market":"cz","worldList":[{"id":"cz28","name":"Bergenhus","active":false,"lastUpdated":1565230980045},{"id":"cz31","name":"Eketorp","active":false,"lastUpdated":1565230980045},{"id":"cz32","name":"Frankenstein","active":false,"lastUpdated":1565230980045},{"id":"cz33","name":"Gaillard","active":false,"lastUpdated":1565230980045},{"id":"cz34","name":"Histria","active":true,"lastUpdated":1567093299284},{"id":"cz35","name":"Inveraray","active":true,"lastUpdated":1567093299284},{"id":"cz36","name":"Jasenov","active":true,"lastUpdated":1567093299284},{"id":"cz37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567093299284},{"id":"cz38","name":"Landskrona","active":true,"lastUpdated":1567093299284}]},{"market":"pt","worldList":[{"id":"pt29","name":"Castelo de Guimarães","active":false,"lastUpdated":1565231022204},{"id":"pt30","name":"Deva","active":false,"lastUpdated":1565231022204},{"id":"pt31","name":"Eketorp","active":false,"lastUpdated":1565231022204},{"id":"pt32","name":"Frankenstein","active":false,"lastUpdated":1565231022204},{"id":"pt33","name":"Gaillard","active":false,"lastUpdated":1567093340318},{"id":"pt34","name":"Histria","active":true,"lastUpdated":1567093340318},{"id":"pt35","name":"Inveraray","active":true,"lastUpdated":1567093340318},{"id":"pt36","name":"Jasenov","active":true,"lastUpdated":1567093340318}]},{"market":"se","worldList":[{"id":"se28","name":"Bergenhus","active":false,"lastUpdated":1557760553881},{"id":"en39","name":"Mosteiro da Batalha","active":false,"lastUpdated":1565231033998},{"id":"en41","name":"Olavinlinna","active":false,"lastUpdated":1565231033998},{"id":"en42","name":"Palamidi","active":false,"lastUpdated":1565231033998},{"id":"en43","name":"Queen's Sconce","active":true,"lastUpdated":1567093360814},{"id":"en44","name":"Rupea","active":true,"lastUpdated":1567093360814},{"id":"en45","name":"Stargard","active":true,"lastUpdated":1567093360814},{"id":"en46","name":"Tzschocha","active":true,"lastUpdated":1567093360814},{"id":"en47","name":"Uhrovec","active":true,"lastUpdated":1567093360814}]},{"market":"no","worldList":[{"id":"no24","name":"Xativa","active":false,"lastUpdated":1557760554264},{"id":"no26","name":"Zipser Burg","active":false,"lastUpdated":1557760554264},{"id":"no27","name":"Alhambra","active":false,"lastUpdated":1557760554264},{"id":"en41","name":"Olavinlinna","active":false,"lastUpdated":1565231122881},{"id":"en39","name":"Mosteiro da Batalha","active":false,"lastUpdated":1565231122881},{"id":"en44","name":"Rupea","active":true,"lastUpdated":1567093443745},{"id":"en43","name":"Queen's Sconce","active":true,"lastUpdated":1567093443745},{"id":"en42","name":"Palamidi","active":false,"lastUpdated":1565231122881},{"id":"en45","name":"Stargard","active":true,"lastUpdated":1567093443745},{"id":"en46","name":"Tzschocha","active":true,"lastUpdated":1567093443745},{"id":"en47","name":"Uhrovec","active":true,"lastUpdated":1567093443745}]},{"market":"sk","worldList":[{"id":"sk30","name":"Deva","active":false,"lastUpdated":1565231307357},{"id":"sk31","name":"Eketorp","active":false,"lastUpdated":1565231307357},{"id":"sk32","name":"Frankenstein","active":false,"lastUpdated":1565231307357},{"id":"sk33","name":"Gaillard","active":false,"lastUpdated":1565231307357},{"id":"sk34","name":"Histria","active":true,"lastUpdated":1567093606307},{"id":"sk35","name":"Inveraray","active":false,"lastUpdated":1567093606307},{"id":"sk36","name":"Jasenov","active":true,"lastUpdated":1567093606307},{"id":"sk37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567093606307}]},{"market":"ro","worldList":[{"id":"ro21","name":"Urquhart Castle","active":false,"lastUpdated":1565231324171},{"id":"ro26","name":"Zipser Burg","active":false,"lastUpdated":1565231324171},{"id":"ro29","name":"Castelo de Guimarães","active":false,"lastUpdated":1565231324171},{"id":"ro30","name":"Deva","active":false,"lastUpdated":1565231324171},{"id":"ro31","name":"Eketorp","active":true,"lastUpdated":1567093620386},{"id":"ro32","name":"Frankenstein","active":false,"lastUpdated":1565231324171},{"id":"ro33","name":"Gaillard","active":true,"lastUpdated":1567093620386},{"id":"ro34","name":"Histria","active":false,"lastUpdated":1567093620386},{"id":"ro35","name":"Inveraray","active":true,"lastUpdated":1567093620386},{"id":"ro36","name":"Jasenov","active":true,"lastUpdated":1567093620386},{"id":"ro37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567093620386}]},{"market":"hu","worldList":[{"id":"hu31","name":"Eketorp","active":false,"lastUpdated":1565231392860},{"id":"hu32","name":"Frankenstein","active":false,"lastUpdated":1565231392860},{"id":"hu33","name":"Gaillard","active":false,"lastUpdated":1565231392860},{"id":"hu34","name":"Histria","active":false,"lastUpdated":1567093651709},{"id":"hu35","name":"Inveraray","active":true,"lastUpdated":1567093651709},{"id":"hu36","name":"Jasenov","active":true,"lastUpdated":1567093651709},{"id":"hu37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567093651709}]},{"market":"br","worldList":[{"id":"br32","name":"Frankenstein","active":false,"lastUpdated":1565231407605},{"id":"br33","name":"Gaillard","active":false,"lastUpdated":1565231407605},{"id":"br34","name":"Histria","active":false,"lastUpdated":1565231407605},{"id":"br35","name":"Inveraray","active":false,"lastUpdated":1565231407605},{"id":"br36","name":"Jasenov","active":true,"lastUpdated":1567093670552},{"id":"br37","name":"Krak des Chevaliers","active":true,"lastUpdated":1567093670552},{"id":"br38","name":"Landskrona","active":true,"lastUpdated":1567093670552},{"id":"br39","name":"Mosteiro da Batalha","active":true,"lastUpdated":1567093670552},{"id":"br40","name":"Navardun","active":true,"lastUpdated":1567093670552}]},{"market":"fi","worldList":[{"id":"fi27","name":"Alhambra","active":false,"lastUpdated":1556219741204},{"id":"fi28","name":"Bergenhus","active":false,"lastUpdated":1556219741204},{"id":"en39","name":"Mosteiro da Batalha","active":false,"lastUpdated":1565231220713},{"id":"en41","name":"Olavinlinna","active":false,"lastUpdated":1565231220713},{"id":"en42","name":"Palamidi","active":false,"lastUpdated":1565231220713},{"id":"en44","name":"Rupea","active":true,"lastUpdated":1567093530218},{"id":"en43","name":"Queen's Sconce","active":true,"lastUpdated":1567093530218},{"id":"en45","name":"Stargard","active":true,"lastUpdated":1567093530218},{"id":"en46","name":"Tzschocha","active":true,"lastUpdated":1567093530218},{"id":"en47","name":"Uhrovec","active":true,"lastUpdated":1567093530218}]}];

/******************************/

type MarketData = {
    market: string,
    worldList: any[]
};

type ComponentProps = RouteComponentProps & Readonly<{
    match: match<{ market: null }>,

    modalSystem: ModalManager,
    featureSettings: ReturnType<typeof featureSettingsInit>,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{
    mapList: MarketData[]
}>;

/******************************/

export default
class PageList
extends React.Component<ComponentProps, ComponentState> {
    // callback do unmount - nasłuch adresu
    unlinstenHistory: Function;

    constructor(props) {
        super(props);

        this.state = {
            mapList: mapListTest
        }

        /******************************/

        this.showSettings = this.showSettings.bind(this);
    }

    /******************************/

    showSettings() {
        this.props.featureSettings('analytics');
    }

    setMap(market: string, worldId: string): void {
        this.props.history.push(`/${ market }/${ worldId }`);
    }
    setMarket(market: string): void {
        this.props.history.push(`/${ market }`);
    }

    /******************************/

    componentDidMount() {
        fetch('/mapList').then((res) => {
            if (res.status === 200) {
                res.json().then(
                    (jsonData) => this.setState({mapList: jsonData}),
                    (err) => null
                );
            }
        });

        this.unlinstenHistory = this.props.history.listen((location, action) => {
            this.props.modalSystem.closeAll();
        });
    }

    componentWillUnmount() {
        this.unlinstenHistory();
    }

    /******************************/

    render(): React.ReactNode {
        let mapList = [],
            { market } = this.props.match.params;

        if (market) {
            // lista światów
            for (let i = 0; i < this.state.mapList.length; i++) {
                let marketData: MarketData = this.state.mapList[ i ],
                    marketKey: string = marketData.market;

                if (market === marketKey) {
                    let worldList: any = marketData.worldList;

                    mapList.push(
                        <div
                            key={'header'}
                            className="d-flex justify-content-between mb-2"
                        >
                            <img className="img-flag" src={ `/flags/${ marketKey }.svg` } />
                            &nbsp;
                            { MARKET_NAME[ marketKey ] || marketKey }
                            <BtnBackToHome
                                history={ this.props.history }
                            />
                        </div>
                    );

                    for (let i = 0; i < worldList.length; i++) {
                        let worldData = worldList[ i ];

                        mapList.push(
                            <World
                                key={ worldData.id }
                                worldId={ worldData.id }
                                worldName={ worldData.name }
                                onClick={ () => this.setMap(market, worldData.id) }
                            />
                        );
                    }
                    break;
                }
            }
        } else {
            mapList.push(
                <div
                    key={'header'}
                    className="d-flex justify-content-between mb-2"
                >
                    Available markets
                </div>
            );

            // lista rynków
            for (let i = 0; i < this.state.mapList.length; i++) {
                let marketData: MarketData = this.state.mapList[ i ],
                    market: string = marketData.market;

                if (market === 'undefined') {
                    continue;
                }

                mapList.push(
                    <Market
                        key={ market }
                        marketId={ market }
                        onClick={
                            () => this.setMarket(market)
                        }
                    />
                );
            }
        }

        return (
            <div className="container position-relative">
                <div style={{
                    position: 'absolute',
                    top: 0,
                    left: 0,
                    width: '100%',
                    textAlign: 'right'
                }}>
                    <button
                        className="btn btn-primary"
                        onClick={ this.showSettings }
                    >
                        <FontAwesomeIcon icon="cog" />
                    </button>
                </div>
                <Header />
                <main className="row">
                    <div className="col-8">
                        <Info />
                    </div>
                    <div className="col-4">
                        <div className="jumbotron p-3">
                            { mapList }
                        </div>
                        <div className="jumbotron p-3">
                            <h4>Do you known?</h4>
                            <p>
                                Inleggal scripts to Tribal Wars 2 can be blockked using two shor lines of code?
                            </p>
                            <p>
                                To <kbd>game_XXXXX.js</kbd> just add the <kbd>(function()&#123;</kbd> to start file and the <kbd>&#125;)();</kbd> to end file.
                            </p>
                        </div>
                    </div>
                </main>
                <Footer />
            </div>
        );
    }
}
