// view/services/crypto.tsx
import * as React from 'react';
import ModalManager, { Modal, Header, Body, Footer, Hook } from "libs/ModalManager";
import LazyPromise from 'libs/LazyPromise';
import * as crypt from 'libs/crypto/ec';
import { ViewContext } from 'view/viewContext';

/******************************/


type ComponentProps = Readonly<{
    context: ViewContext,
    hook: Hook,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{
    area1: string,
    area2: string,
    area3: string,
    area4: string
}>

/******************************/

class ModalBody
extends React.Component<ComponentProps, ComponentState> {
    state = {
        area1: '',
        area2: '',
        area3: '',
        area4: ''
    };

    /******************************/

    createKeys = () => {
        const [ privateKey, publicKey ] = crypt.getKeys();

        this.setState({
            area1: privateKey,
            area2: publicKey
        });
    }

    createSignature = async () => {
        const sign = crypt.sign(
            this.state.area1,
            this.state.area3
        );

        this.setState({
            area4: sign
        });
    }

    testSignature = () => {
        const result = crypt.verify(
            this.state.area2,
            this.state.area3,
            this.state.area4
        );

        this.setState({
            area4: result ? 'true' : 'false'
        });
    }

    /******************************/

    render(): JSX.Element {
        return(
            <div className="px-3 py-1">
                <h6>Private key</h6>
                <textarea
                    className="my-2"
                    cols={ 33 }
                    rows={ 3 }
                    onChange={(e) => this.setState({area1: e.target.value})}
                    value={ this.state.area1 }
                ></textarea>
                <h6>Public key</h6>
                <textarea
                    className="my-2"
                    cols={ 33 }
                    rows={ 3 }
                    onChange={(e) => this.setState({area2: e.target.value})}
                    value={ this.state.area2 }
                ></textarea>
                <h6>Value</h6>
                <textarea
                    className="my-2"
                    cols={ 33 }
                    rows={ 3 }
                    onChange={(e) => this.setState({area3: e.target.value})}
                    value={ this.state.area3 }
                ></textarea>
                <h6>Signature</h6>
                <textarea
                    className="my-2"
                    cols={ 33 }
                    rows={ 3 }
                    onChange={(e) => this.setState({area4: e.target.value})}
                    value={ this.state.area4 }
                ></textarea>
                <div className="btn-group">
                    <div
                        className="btn btn-primary"
                        onClick={ this.createKeys }
                    >
                        Create keys
                    </div>
                </div>
                <div className="btn-group">
                    <div
                        className="btn btn-primary"
                        onClick={ this.createSignature }
                    >
                        Sign
                    </div>
                    <div
                        className="btn btn-primary"
                        onClick={ this.testSignature }
                    >
                        Verify
                    </div>
                </div>
            </div>
        );
    }
}

/******************************/

export default function serviceCrypto(
    refContext: any
) {
    const context = refContext as ViewContext

    function openModal(): Promise<string> {
        let hook: Hook = context.modalSystem.createHook(),
            lazyPromise = new LazyPromise<string>();

        function onClose(): void {
            hook.close();
            lazyPromise.reject(null);
        }

        hook.setLayout(
            <Modal onClose={ onClose }>
                <Header>Crypto</Header>
                <Body className="p-1">
                    <ModalBody context={ context } hook={ hook } />
                </Body>
            </Modal>
        );
        hook.setCursorPosition();

        return lazyPromise.get();
    }

    return {
        openModal
    };
}
export type Prompt = ReturnType<typeof serviceCrypto>;
