// view/components/search/ResultItem.tsx
import * as React from 'react';
import { Api } from 'core';
import Item from 'core/data/objects/Item';

/******************************/

const stopPropagation = (e) => {
    e.stopPropagation();
};

/******************************/

type ComponentProps = Readonly<{
    item: Item,
    map: Api,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{}>

/******************************/

export default
abstract class ResultItem
extends React.Component<ComponentProps, ComponentState> {
    constructor(props) {
        super(props);

        /******************************/

        this.onMouseEnter = this.onMouseEnter.bind(this);
        this.onMouseLeave = this.onMouseLeave.bind(this);
    }

    /******************************/

    onMouseEnter(e: any): void {
        this.props.map.events.emit('hover/item', this.props.item);
    }
    onMouseLeave(e: any): void {
        this.props.map.events.emit('hover/item', null);
    }

    /******************************/

    getButtons(): void {
        return null;
    }

    abstract getIcon(): JSX.Element

    render(): JSX.Element {
        return (
            <div
                className="item px-2 py-2"
                onMouseEnter={ this.onMouseEnter }
                onMouseLeave={ this.onMouseLeave }
            >
                <div className="text-nowrap">
                    <span className="icon px-1">
                        { this.getIcon() }
                    </span>
                    <span className="name px-1">
                        { this.props.item.name }
                    </span>
                </div>
                { this.getButtons() }
            </div>
        );
    }
}
