// view/components/search/ResultItem.tsx
import * as React from 'react';
import Item from 'core/data/objects/Item';

/******************************/

const stopPropagation = (e) => {
    e.stopPropagation();
};

/******************************/

type ComponentProps = Readonly<{
    item: Item,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{}>

/******************************/

export default
abstract class ResultItem
extends React.Component<ComponentProps, ComponentState> {
    constructor(props) {
        super(props);

        /******************************/

        this.onClick = this.onClick.bind(this);
        this.onMouseMove = this.onMouseMove.bind(this);
        this.onMouseLeave = this.onMouseLeave.bind(this);
    }

    /******************************/

    onClick(e: any): void {
        e.stopPropagation();
    }

    onMouseMove(e: any): void {
    }

    onMouseLeave(e: any): void {
    }

    /******************************/

    getButtons(): void {
        return null;
    }

    abstract getIcon(): JSX.Element

    render(): JSX.Element {
        return (
            <div
                className="item"
                onClick={ this.onClick }
                onMouseMove={ this.onMouseMove }
                onMouseDown={ stopPropagation }
                onMouseLeave={ this.onMouseLeave }
            >
                <div className="text-nowrap">
                    <span className="icon">
                        { this.getIcon() }
                    </span>
                    <span className="name">
                        { this.props.item.name }
                    </span>
                </div>
                { this.getButtons() }
            </div>
        );
    }
}
