// libs/ModalManager/types.ts
import * as React from 'react';

export type Position = {
    x: number,
    y: number
};

export type Size = {
    w: number,
    h: number
}

// Hook

export enum Orientation {
    TL = 0,
    TR = 1,
    BL = 2,
    BR = 3
}

export type HookOptionsTemplateItemTypes = JSX.Element | boolean | null;
export type HookOptionsTemplateItemBuilderTemplate = () => HookOptionsTemplateItemTypes;
export type HookOptionsTemplateItem = HookOptionsTemplateItemTypes | HookOptionsTemplateItemBuilderTemplate;

export type HookOptionsTemplateTypes = JSX.Element | null;
export type HookOptionsBuilderTemplate = () => HookOptionsTemplateTypes;
export type HookOptionsTemplate = HookOptionsTemplateTypes | HookOptionsBuilderTemplate;

export type HookOptionsTemplateContentTypes = string | JSX.Element | boolean | null;
export type HookOptionsBuilderTemplateContent = () => HookOptionsTemplateContentTypes;
export type HookOptionsTemplateContent = HookOptionsTemplateContentTypes | HookOptionsBuilderTemplateContent;

export type HookOptions = {
    hidden?: boolean, // po zamontowaniu ma być widoczny/ukryty
    alwaysMounted?: boolean,
    position?: Position,
    autoPosition?: boolean,
    autoMount?: boolean,
    spyCursor?: boolean,
    hideWhenInactive?: boolean,

    background?: HookOptionsTemplateItem,
    header?: HookOptionsTemplateContent,
    headerTemplate?: HookOptionsTemplate,
    closeButton?: HookOptionsTemplateItem,
    body?: HookOptionsTemplateContent,
    bodyTemplate?: HookOptionsTemplate,
    footer?: HookOptionsTemplateContent,
    footerTemplate?: HookOptionsTemplate
    style?: React.CSSProperties
};
