// view/components/settings/MapColors.tsx
import * as React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { CanvasApi } from 'canvas';
import { ColorPicker } from 'view/services/serviceColorPicker';
import { StoreHelper } from 'stores';
import { WorldStore } from 'stores/world';
import { GlobalStore } from 'stores/global';
import * as mapColors from 'stores/@/MapColors';

/******************************/

type ComponentProps = Readonly<{
    store: WorldStore | GlobalStore,
    storeHelper: StoreHelper,
    mainCanvas: CanvasApi,
    serviceColorPicker: ColorPicker,
    allowDisable?: boolean,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{}>

/******************************/

export default
class MapColors
extends React.Component<ComponentProps, ComponentState> {
    unsubscriberList = [];

    componentDidMount(): void {
        this.unsubscriberList.push(
            this.props.store.subscribe(
                () => this.forceUpdate()
            )
        );
    }

    componentWillUnmount(): void {
        let unsubscriber;

        while (unsubscriber = this.unsubscriberList.pop()) {
            unsubscriber();
        }
    }

    /******************************/

    render(): JSX.Element {
        let setColor = this.props.mainCanvas.commands.setColorInSlot,
            store = this.props.store,
            state = store.getState(),
            cPick = this.props.serviceColorPicker,
            mergedColors = this.props.storeHelper.getMapColors(),
            currentColors = state.mapColors;

        let list = [];
        for (let index in mapColors.maps.index2color) {
            let slot = parseInt(index),
                { key, label } = mapColors.maps.index2color[ index ],
                currentColor = currentColors[ key ],
                previousColor = mergedColors[ key ],
                btnOffClass = currentColor ? 'btn-success' : 'btn-dark'

            list.push(
                <div key={ key } className="form-group m-0 d-flex justify-content-between">
                    <label className="my-2">{ label }</label>
                    <div className="btn-group my-2 fx-ac">
                        <button
                            className="btn btn-dark btn-sm p-3 mx-2"
                            style={{
                                backgroundColor: currentColor || previousColor
                            }}
                            onClick={(e) => {
                                cPick(
                                    currentColor || previousColor,
                                    (color) => setColor(slot, color)
                                ).then(
                                    (color) => {
                                        setColor(slot, previousColor);
                                        store.actions.setMapColor(key, color);
                                        this.forceUpdate();
                                    },
                                    (error) => setColor(slot, previousColor)
                                );
                            }}
                        ></button>
                        {
                            !this.props.allowDisable ? null :
                            <button
                                className={`btn ${ btnOffClass } px-2 py-1`}
                                onClick={(e) => {
                                    store.actions.setMapColor(key, null);
                                    this.forceUpdate();
                                }}
                            >
                                <FontAwesomeIcon icon="power-off" />
                            </button>
                        }
                    </div>
                </div>
            );
        }

        return (
            <div className="ofh">{ list }</div>
        );
    }
}
