// view/components/settings/MapSettings.tsx
import * as React from 'react';
import { Api } from 'core';
import { StoreMap } from 'stores';
import serviceColorPickerInit from 'view/services/serviceColorPicker';

import { map_index2color } from 'data/mapColors';

/******************************/

type ComponentProps = Readonly<{
    stores: StoreMap,
    mainMap: Api,
    serviceColorPicker: ReturnType<typeof serviceColorPickerInit>,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{}>

/******************************/

export default
class MapSettings
extends React.Component<ComponentProps, ComponentState> {
    unsubscriberList = [];

    constructor(props) {
        super(props);
    }

    /******************************/

    componentWillMount(): void {
        this.unsubscriberList.push(
            this.props.stores.global.subscribe(
                () => this.forceUpdate()
            )
        );
    }

    componentWillUnmount(): void {
        for (let unsubscriber of this.unsubscriberList) {
            unsubscriber();
        }
    }

    /******************************/

    render(): JSX.Element {
        let bM = this.props.mainMap.buffersManager,
            globalStore = this.props.stores.global,
            globalState = globalStore.getState(),
            setColor = this.props.mainMap.helpers.setColor,
            cPick = this.props.serviceColorPicker;

        let list = [];
        for (let index in map_index2color) {
            let slot = parseInt(index),
                { key, label } = map_index2color[ index ];

            list.push(
                <div key={ key } className="form-group m-0 d-flex justify-content-between">
                    <label className="my-2">{ label }</label>
                    <div className="my-2 fx-ac">
                        <button
                            className="btn btn-dark btn-sm p-3 mx-2"
                            style={{
                                backgroundColor: globalState[ `map_${ key }` ]
                            }}
                            onClick={(e) => {
                                cPick(
                                    globalState[ `map_${ key }` ],
                                    (color) => setColor(slot, color)
                                ).then(
                                    (color) => {
                                        globalStore.actions.setMapColor(key, color);
                                        this.forceUpdate();
                                    },
                                    (error) => {
                                        setColor(slot, globalState[ `map_${ key }` ]);
                                    }
                                );
                            }}
                        ></button>
                    </div>
                </div>
            );
        }

        return (
            <div className="ofh">{ list }</div>
        );
    }
}
