// view/components/settings/PrivacySettings.tsx
import * as React from 'react';
import ButtonOnOff from 'view/components/ButtonOnOff';
import { ViewContext } from 'view/viewContext';

/******************************/

type ComponentProps = Readonly<{
    children?: React.ReactNode,
    context: ViewContext
}>;
type ComponentState = Readonly<{}>

/******************************/

export default
class PrivacySettings
extends React.Component<ComponentProps, ComponentState> {
    constructor(props) {
        super(props);
    }

    /******************************/

    unsubscribe_list: Array<() => void> = [];

    componentDidMount() {
        const store = this.props.context.stores.global;

        this.unsubscribe_list.push(
            store.subscribe(
                () => this.forceUpdate()
            )
        );
    }

    componentWillUnmount() {
        for (let i = 0; i < this.unsubscribe_list.length; i++) {
            this.unsubscribe_list[ i ]();
        }
    }

    /******************************/

    render(): React.ReactNode {
        const store = this.props.context.stores.global;
        const dontSpyAction = store.actions.setDontSpy as (value: boolean) => void;
        const state = store.getState();

        return (
            <table className="table privacy-settings mb-0">
                <thead>
                    <tr>
                        <th>Service</th>
                        <th>Description</th>
                        <th>Options</th>
                    </tr>
                </thead>
                <tbody>
                    <tr>
                        <td className="p-2">
                            Google Analytics
                        </td>
                        <td className="p-2">
                            When active, it's your moves on&nbsp;this page as&nbsp;an&nbsp;one user. That's collect informations as: an&nbsp;url vised pages, an&nbsp;url provious site, your's IP&nbsp;address, a&nbsp;language settings, a&nbsp;charset settings. In&nbsp;eles, that's collect an&nbsp;information about views pages.
                        </td>
                        <td className="p-2">
                            <ButtonOnOff
                                value={ !state.analytics.dontSpy }
                                onChange={
                                    (value) => dontSpyAction(!value)
                                }
                            />
                        </td>
                    </tr>
                </tbody>
            </table>
        );
    }
}
