// view/components/settings/SettingsBody.tsx
import * as React from 'react';
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
import MapColors from 'view/components/settings/MapColors';
import { ViewContext } from 'view/viewContext';
import { Hook } from 'libs/ModalManager';
import PrivacySettings from 'view/components/settings/PrivacySettings';
import * as $store from 'stores/state';

/******************************/

type ComponentProps = Readonly<{
    hook: Hook,
    context: ViewContext,
    children?: React.ReactNode,
}>;
type ComponentState = Readonly<{
    storeState: $store.State
}>

/******************************/

export default
class SettingsBody
extends React.Component<ComponentProps, ComponentState> {
    state = {storeState: null};

    /******************************/

    unsubscriberList = [];

    componentWillUnmount(): void {
        let unsubscriber;

        while (unsubscriber = this.unsubscriberList.pop()) {
            unsubscriber();
        }
    }

    /******************************/

    getState(): $store.State {
        if (this.state.storeState !== null) {
            return this.state.storeState;
        }

        const store = this.props.context.stores.state;

        this.unsubscriberList.push(
            store.subscribe(() => this.setState({storeState: store.getState()}))
        );

        return store.getState();
    }

    /******************************/

    render(): React.ReactNode {
        const { hook, context } = this.props;
        const state = this.getState();

        return (
            <Tabs
                onSelect={ hook.update }
            >
                <TabList>
                    <Tab>Global</Tab>
                    { state.location.worldKey
                        ? <Tab>World</Tab>
                        : null
                    }
                </TabList>

                <TabPanel>
                    <Tabs
                        onSelect={ hook.update }
                    >
                        <TabList>
                            <Tab>Map colors</Tab>
                            <Tab>Privacy settings</Tab>
                        </TabList>

                        <TabPanel>
                            <MapColors
                                context={ context }
                                store={ context.stores.global }
                            />
                        </TabPanel>
                        <TabPanel>
                            <PrivacySettings context={ context } />
                        </TabPanel>
                    </Tabs>
                </TabPanel>


                { state.location.worldKey
                    ? (
                        <TabPanel>
                            <Tabs
                                onSelect={ hook.update }
                            >
                                <TabList>
                                    <Tab>Map colors</Tab>
                                </TabList>

                                <TabPanel>
                                    <MapColors
                                        allowDisable={ true }
                                        context={ context }
                                        store={ context.stores.world }
                                    />
                                </TabPanel>
                            </Tabs>
                        </TabPanel>
                    )
                    : null
                }
            </Tabs>
        );
    }
}
