// core/index.ts
import Renderer from 'core/Renderer';
import BuffersManager from 'core/BuffersManager';
import { dataLoader_v1, bitDataLoader_v1 } from 'core/data/loader';
import MapData from 'core/data/MapData';
import State from 'core/State';
import { initMouseEvents, initResize } from 'core/eventService';
import EventBox from 'core/EventBox';

/******************************/

export type Api = {
    canvas: HTMLCanvasElement,
    state: State,
    buffersManager: BuffersManager,
    events: EventBox,
    initPromise: Promise<void>
}

/******************************/

export default function initMap(
    canvas: HTMLCanvasElement
): Api {
    let events = new EventBox();
    let buffersManager = new BuffersManager();
    let mapData = new MapData();
    let renderer = new Renderer(canvas);
    let state = new State();
    let initPromise = (async function(){
        buffersManager.init();
        await renderer.init(
            buffersManager.getDataBuffer(),
            buffersManager.getColorBuffer()
        );

        initMouseEvents(canvas, renderer, state, events);
        initResize(canvas, renderer, state);

        await bitDataLoader_v1(
            'data/mapv2-rc1.bin',
            buffersManager
        );
        renderer.updateData(buffersManager.getDataBuffer());
        renderer.requestUpdate();

        await dataLoader_v1(
            'data/data.json',
            mapData,
            buffersManager
        );
        renderer.updateData(buffersManager.getDataBuffer());
        renderer.requestUpdate();
    })();

    return {
        state, buffersManager, events,
        initPromise, canvas
    }
}
