// helpers/time.ts
export function addLendingZeros(
    number: number,
    length: number
): string {
    let str: string = '' + number;

    while (str.length < length) {
        str = '0' + str;
    }

    return str;
}

export function timeToSQLDateTime(
    time: number,
    noSeconds: boolean
): string {
    var date = new Date(time),
        str = date.getFullYear() +
            '-' +
            addLendingZeros(date.getMonth() + 1, 2) +
            '-' +
            addLendingZeros(date.getDate(), 2) +
            ' ' +
            addLendingZeros(date.getHours(), 2) +
            ':' +
            addLendingZeros(date.getMinutes(), 2);

    if (!noSeconds) {
        str+= ':' + addLendingZeros(date.getSeconds(), 2);
    }

    return str;
}

export function parseSecond(
    s: number
): string {
    let temp: number,
        str: string,
        ss: number;

    temp = s % 60;
    str = temp +'s';
    ss = (s - temp) / 60;

    if (!s) {
        return str;
    }

    temp = ss % 60;
    str = temp +'m '+ str;
    ss = (ss - temp) / 60;

    if (!ss) {
        return str;
    }

    temp = ss % 24;
    str = temp +'h '+ str;
    ss = (ss - temp) / 24;

    if (!ss) {
        return str;
    }

    str = ss +'d '+ str;

    return str;
}
