// stores/@/MapColors.ts
export const COLOR_INDEX_BACKGROUND = 0;
export const COLOR_INDEX_BORDER_PROVINCE = 1;
export const COLOR_INDEX_BORDER_CONTINENT = 2;
export const COLOR_INDEX_VILLAGE_PLAYER = 3;
export const COLOR_INDEX_VILLAGE_BARBARIAN = 4;
export const COLOR_INDEX_HOVER_ITEM = 5;
export const COLOR_INDEX_HOVER_GROUP = 6;

/******************************/

export default class MapColors {
    background = {active: false, value: '#004000'};
    borderProvince = {active: false, value: '#888888'};
    borderContinent = {active: false, value: '#ffff00'};
    villagePlayer = {active: false, value: '#003000'};
    villageBarbarian = {active: false, value: '#005000'};
    hoverItem = {active: false, value: '#ffffff'};
    hoverGroup = {active: false, value: '#999999'};
}

type MapColorItem = {
    index: number,
    label: string,
    key: keyof MapColors,
    initValue: string
};

export const mapColorList: Readonly<MapColorItem[]> = [
    {
        index: COLOR_INDEX_BACKGROUND,
        label: 'Background',
        key: 'background',
        initValue: '#004000'
    },
    {
        index: COLOR_INDEX_BORDER_PROVINCE,
        label: 'Province border',
        key: 'borderProvince',
        initValue: '#888888'
    },
    {
        index: COLOR_INDEX_BORDER_CONTINENT,
        label: 'Continent border',
        key: 'borderContinent',
        initValue: '#ffff00'
    },
    {
        index: COLOR_INDEX_VILLAGE_PLAYER,
        label: 'Player\'s villages',
        key: 'villagePlayer',
        initValue: '#003000'
    },
    {
        index: COLOR_INDEX_VILLAGE_BARBARIAN,
        label: 'Barbarian\'s villages',
        key: 'villageBarbarian',
        initValue: '#005000'
    },
    {
        index: COLOR_INDEX_HOVER_ITEM,
        label: 'Hover of item',
        key: 'hoverItem',
        initValue: '#ffffff'
    },
    {
        index: COLOR_INDEX_HOVER_GROUP,
        label: 'Hover of group',
        key: 'hoverGroup',
        initValue: '#999999'
    }
];

const map_index2color: {[key: number]: MapColorItem} = {};
for (let item of mapColorList) {
    map_index2color[ item.index ] = item;
}

//@ts-ignore
const map_key2color: {[key in keyof MapColors]: MapColorItem} = {};
for (let item of mapColorList) {
    map_key2color[ item.key ] = item;
}

export const maps = {
    index2color: map_index2color,
    key2color: map_key2color
}

/******************************/

type ColorData = {
    key: keyof MapColors,
    value: {
        active: boolean
        color: string
    }
};

//@ts-ignore
export const initStateMapColors: {[key in keyof MapColors]: {active: boolean, value: string | null}} = {};

mapColorList.map((item) => {
    initStateMapColors[ item.key ] = {active: false, value: item.initValue};
});

export function createReducers<State extends {mapColors: {[key in keyof MapColors]: {active: boolean, value: string | null}}}>() {
    return {
        'mapColors/setColor': (state: State, value: ColorData) => {
            const nextState = { ...state };
            const nextStateMapColors = { ...nextState.mapColors };

            nextStateMapColors[ value.key ] = {active: value.value.active, value: value.value.color};
            nextState.mapColors = nextStateMapColors;

            return nextState;
        }
    };
};

export function createActions<Store>(this: Store) {
    return {
        setMapColor(
            key: keyof MapColors,
            value: ColorData['value']
        ): void {
            this.dispatch({
                type: 'mapColors/setColor',
                value: { key, value }
            });
        }
    };
};
