// canvas/commands/goTo.ts
import State from "canvas/data/State";
import { Commands } from "canvas";

/******************************/

export default function commandGoTo(
    state: State,
    cmd: any
) {
    const commands = cmd as Commands;

    let startPosition = [0, 0],
        endPosition = [0, 0],
        startTime = 0,
        endTime = 0;

    function animatePosition(): void {
        let now = Date.now(),
            progress: number;

        if (now < endTime) {
            requestAnimationFrame(animatePosition);
            progress = 1 - (endTime - now) / (endTime - startTime);
            progress = 1 - 1 / (1 + Math.pow(Math.E, ((progress - 0.5)*12)));
        } else {
            progress = 1;
        }

        commands.setPosition(
            startPosition[0] + (endPosition[0] - startPosition[0]) * progress,
            startPosition[1] + (endPosition[1] - startPosition[1]) * progress
        );
    }

    return {
        goTo(
            x: number,
            y: number
        ): void {
            startPosition = state.getPosition();
            endPosition = state.calcTile2Value(x, y);

            startTime = Date.now();
            endTime = startTime + 5e2;

            requestAnimationFrame(animatePosition);
        }
    };
}
