// canvas/commands/setHover.ts
import Events from "canvas/Events";
import State from "canvas/data/State";
import { Commands } from "canvas";
import Item from "data/data/Item";

/******************************/

export default function commandSetHover(
    events: Events,
    state: State,
    cmd
) {
    let commands = cmd as Commands;

    return {
        setHoverTile(
            tileX: number,
            tileY: number
        ): void {
            let [ currentTileX, currentTileY ] = state.getHoverTile();

            if (currentTileX !== tileX
                || currentTileY !== tileY
            ) {
                state.setHoverTile(tileX, tileY);
                events.emit('changed hover/tile', { x: tileX, y: tileY });
            }
        },
        setHoverScreenPx(
            pxX: number,
            pxY: number
        ): void {
            let [ halfWidth, halfHeight ] = state.getScreenSizeHalf(),
                [ posPxX, posPxY ] = state.getPositionPixel(),
                [ tileX, tileY ] = state.calcPixel2Tile(
                    posPxX - halfWidth + pxX,
                    posPxY - halfHeight + pxY
                );

            state.setHoverScreenPx(pxX, pxY);
            commands.setHoverTile(tileX, tileY);
        },
        setHoverItem(item: Item): void {
            let hoveredItem = state.getHoverItem();

            if (hoveredItem !== item) {
                state.setHoverItem(item);
                events.emit('changed hover/item', item);
            }
        }
    };

    // hover item on map

}
