// canvas/commands/setScreenSize.ts
import State from "canvas/data/State";
import Renderer from "canvas/Renderer";
import { Commands } from "canvas";

/******************************/

export default function commandSetScreenSize(
    canvas: HTMLCanvasElement,
    state: State,
    renderer: Renderer,
    cmd: any
) {
    const commands = cmd as Commands;

    return {
        setScreenSize(
            w: number,
            h: number
        ): void {
            state.setScreenSize(w, h);

            commands.setPosition(
                ...state.getPosition()
            );

            renderer.updateResolution();
        },
        setParentElementSize(): void {
            if (canvas.parentElement) {
                let { clientWidth: width, clientHeight: height } = canvas.parentElement;

                canvas.setAttribute('width', width.toString());
                canvas.setAttribute('height', height.toString());

                Object.assign(canvas.style, {
                    width: `${ width }px`,
                    height: `${ height }px`
                });

                commands.setScreenSize(width, height);
            }
        }
    };
}
