// view/features/modalSearch.tsx
import * as React from 'react';
import ModalManager, { Modal, Header, Body, Footer, Hook } from "libs/ModalManager";
import { CanvasApi } from 'canvas';
import LazyPromise from 'libs/LazyPromise';
import Search from 'view/components/search/Search';
import { MapDataApi } from 'data';

/******************************/

export default function featureSearch(
    modalManager: ModalManager,
    canvasApi: CanvasApi,
    appData: MapDataApi
): (tab?: string) => Promise<string> {
    let hook: Hook = modalManager.createHook(),
        lazyPromise = new LazyPromise();

    hook.setLayout(
        <Modal
            hidden={ true }
            hideNotUnmount={ true }
            className="searchModal"
        >
            <Header>
                Search
            </Header>
            <Body className="p-2">
                <Search
                    hook={ hook }
                    canvasApi={ canvasApi }
                    search={ appData.commands.search }
                    updateModal={ hook.update }
                />
            </Body>
        </Modal>
    );

    function open(): Promise<string> {
        if (hook.isHidden()) {
            hook.show();
            hook.setCursorPosition();
        }

        return lazyPromise.get();
    }

    window['mSearch'] = hook;

    return open;
}
