// view/features/modalSettings.tsx
import * as React from 'react';
import ModalManager, { Modal, Header, Body, Footer, Hook } from "libs/ModalManager";
import { CanvasApi } from 'canvas';
import createStores from 'stores';
import serviceColorPickerInit from 'view/services/serviceColorPicker';
import LazyPromise from 'libs/LazyPromise';

import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
import MapSettings from 'view/components/settings/MapSettings';

/******************************/

export default function featureSettings(
    modalManager: ModalManager,
    stores: ReturnType<typeof createStores>,
    mainCanvas: CanvasApi,
    serviceColorPicker: ReturnType<typeof serviceColorPickerInit>
): (tab?: string) => Promise<string> {
    let hook: Hook = modalManager.createHook(),
        lazyPromise = new LazyPromise();

    hook.setLayout(
        <Modal
            hidden={ true }
            hideNotUnmount={ true }
        >
            <Header>
                Settings
            </Header>
            <Body style={{ minWidth: '20em' }}>
                <Tabs>
                    <TabList>
                        <Tab>Map settings</Tab>
                    </TabList>

                    <TabPanel>
                        <MapSettings
                            mainCanvas={ mainCanvas }
                            stores={ stores }
                            serviceColorPicker={ serviceColorPicker }
                        />
                    </TabPanel>
                </Tabs>
            </Body>
        </Modal>
    );

    function open(
        tab?: string
    ): Promise<string> {
        if (hook.isHidden()) {
            hook.show();
            hook.setCursorPosition();
        }

        return lazyPromise.get();
    }

    window['mSettings'] = hook;

    return open;
}
