// view/features/modalSettings.tsx
import * as React from 'react';
import ModalManager, { Modal, Header, Body, Footer, Hook } from "libs/ModalManager";
import { CanvasApi } from 'canvas';
import { StoreBox } from 'stores';
import serviceColorPickerInit from 'view/services/serviceColorPicker';
import LazyPromise from 'libs/LazyPromise';

import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
import MapColors from 'view/components/settings/MapColors';

/******************************/

export default function featureSettings(
    modalManager: ModalManager,
    stores: StoreBox,
    mainCanvas: CanvasApi,
    serviceColorPicker: ReturnType<typeof serviceColorPickerInit>
): (tab?: string) => Promise<string> {
    let hook: Hook = modalManager.createHook(),
        lazyPromise = new LazyPromise();

    hook.setLayout(
        <Modal
            hidden={ true }
            hideNotUnmount={ true }
        >
            <Header>
                Settings
            </Header>
            <Body style={{ minWidth: '20em' }}>
                <Tabs>
                    <TabList>
                        <Tab>World</Tab>
                        <Tab>Global</Tab>
                    </TabList>

                    <TabPanel>
                        <Tabs>
                            <TabList>
                                <Tab>Map colors</Tab>
                            </TabList>

                            <TabPanel>
                                <MapColors
                                    allowDisable={ true }
                                    mainCanvas={ mainCanvas }
                                    store={ stores.world }
                                    storeHelper={ stores.helper }
                                    serviceColorPicker={ serviceColorPicker }
                                />
                            </TabPanel>
                        </Tabs>
                    </TabPanel>

                    <TabPanel>
                        <Tabs>
                            <TabList>
                                <Tab>Map colors</Tab>
                            </TabList>

                            <TabPanel>
                                <MapColors
                                    mainCanvas={ mainCanvas }
                                    store={ stores.global }
                                    storeHelper={ stores.helper }
                                    serviceColorPicker={ serviceColorPicker }
                                />
                            </TabPanel>
                        </Tabs>
                    </TabPanel>
                </Tabs>
            </Body>
        </Modal>
    );

    function open(
        tab?: string
    ): Promise<string> {
        if (hook.isHidden()) {
            hook.show();
            hook.setCursorPosition();
        }

        return lazyPromise.get();
    }

    window['mSettings'] = hook;

    return open;
}
