// view/features/modalMarkers.tsx
import * as React from 'react';
import ModalManager, { Modal, Header, Body, Footer, Hook } from "libs/ModalManager";
import { CanvasApi } from 'canvas';
import { StoreBox } from 'stores';
import LazyPromise from 'libs/LazyPromise';
import Markers from 'view/components/markers/Markers';

/******************************/

export default function featureMarkers(
    modalManager: ModalManager,
    canvasApi: CanvasApi,
    stores: StoreBox
): () => Promise<string> {
    let hook: Hook = modalManager.createHook(),
        lazyPromise = new LazyPromise();

    hook.setLayout(
        <Modal
            hidden={ true }
            hideNotUnmount={ true }
            className="markersModal"
        >
            <Header>
                Markers
            </Header>
            <Body className="p-2">
                <Markers
                    hook={ hook }
                    canvasApi={ canvasApi }
                    stores={ stores }
                    updateModal={ hook.update }
                />
            </Body>
        </Modal>
    );

    function open(): Promise<string> {
        if (hook.isHidden()) {
            hook.show();
            hook.setCursorPosition();
        }

        return lazyPromise.get();
    }

    window['mMarkers'] = hook;

    return open;
}
