// view/components/markers/ModalBody.tsx
import * as React from 'react';
import debug from 'debug';
import { Hook } from 'libs/ModalManager';
import { ViewContext } from 'view/viewContext';
import Marker from 'view/components/markers/Marker';
import { IconOnOff } from 'view/components/icon';

/******************************/

const debugLog = debug('app:view:markers:modalBody');

/******************************/

type ComponentProps = Readonly<{
    hook: Hook,
    context: ViewContext,
    updateModal: () => void,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{}>

/******************************/

export default
class MarkersModalBody
extends React.Component<
    ComponentProps,
    ComponentState
> {
    unsubscriberList = [];
    forceUpdate = () => null

    componentWillMount(): void {
        const context = this.props.context;
        let previousStore = null;

        this.unsubscriberList.push(
            context.stores.world.subscribe(
                () => {
                    const worldState = context.stores.world.getState();

                    if (previousStore === null
                        || worldState.markers !== previousStore.markers
                    ) {
                        this.forceUpdate();
                        previousStore = worldState;
                    }
                }
            ),
            context.dragDrop.subscribe(
                () => this.forceUpdate()
            )
        );
        this.forceUpdate();

        this.props.context.data.events.once('changed data/base', () => this.forceUpdate());
    }

    componentDidMount(): void {
        this.forceUpdate = (...args) => super.forceUpdate.apply(this, args);
    }

    componentWillUnmount(): void {
        let unsubscriber;

        while (unsubscriber = this.unsubscriberList.pop()) {
            unsubscriber();
        }

        this.forceUpdate = () => null;
    }

    componentDidUpdate(): void {
        this.props.updateModal();
    }

    /******************************/

    render(): JSX.Element {
        debugLog('render');
        const { context } = this.props;
        const cPick = context.services.colorPicker;
        const store = context.stores.world;

        let worldState = this.props.context.stores.world.getState();
        let markersState = worldState.getMarkers();
        let activePackage = markersState.getActivePackage();
        let rootMarker = activePackage.getRootMarker();
        let btnOffClass = rootMarker.colorActive ? 'btn-success' : 'btn-dark';

        const onClickBtnColorActive = () => {
            store.actions.markerSetColorActive(
                rootMarker.uuid,
                !rootMarker.colorActive
            );
        };

        return (
            <>
                <div className="marker active">
                    <div className="marker-header value-format">
                        <div>
                            <span className="unit">Package:</span>
                            <span className="value-0">
                                { activePackage.name }
                            </span>
                        </div>
                        <div className="btn-group ifx-as">
                            <button
                                className={`btn ${ btnOffClass } px-5`}
                                style={{
                                    'backgroundColor': rootMarker.colorValue
                                }}
                                onClick={(e) => {
                                    cPick(
                                        rootMarker.colorValue,
                                        (color) => {}
                                    ).then(
                                        (color) => {
                                            store.actions.markerSetColor(rootMarker.uuid, color);
                                        },
                                        (error) => {}
                                    );
                                }}
                            ></button>
                            <button
                                className={`btn ${ btnOffClass } btn-xs px-3`}
                                onClick={ onClickBtnColorActive }
                            >
                                <IconOnOff />
                            </button>
                        </div>
                    </div>
                </div>

                <Marker
                    context={ this.props.context }
                    marker={ rootMarker }
                    index={ 0 }
                    package={ activePackage }
                />
            </>
        );
    }
}
