// view/components/topBar/BtnStats.tsx
import * as React from 'react';
import Events from 'map/Events';

/******************************/

type ComponentProps = Readonly<{
    eventCore: Events,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{
    currentCoords: {
        x: number,
        y: number
    }
}>

/******************************/

export default
class BtnStats
extends React.Component<ComponentProps, ComponentState> {
    constructor(props) {
        super(props);

        this.state = {
            currentCoords: {
                x: 0,
                y: 0
            }
        }
    }

    /******************************/

    componentDidMount(): void {
        this.props.eventCore.on('hover/map', (data) => {
            this.setState({
                currentCoords: {
                    x: data.x,
                    y: data.y
                }
            });
        });
    }

    /******************************/

    render(): JSX.Element {
        return (
            <div className="btn-group mx-1 ifx-as">
                <button
                    className="btn btn-primary value-format ifx-ac"
                >
                    <span className="value-1">
                        { this.state.currentCoords.x }
                    </span>
                    <span className="special-char">|</span>
                    <span className="value-1">
                        { this.state.currentCoords.y }
                    </span>
                </button>
            </div>
        );
    }
}
