// view/components/topBar/BtnsStats.tsx
import * as React from 'react';
import Events from 'core/Events';

/******************************/

type ComponentProps = Readonly<{
    eventCore: Events,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{
    currentCoords: {
        x: number,
        y: number
    }
}>

/******************************/

export default
class BtnsStats
extends React.Component<ComponentProps, ComponentState> {

    constructor(props) {
        super(props);

        this.state = {
            currentCoords: {
                x: 0,
                y: 0
            }
        }
    }

    /******************************/

    componentDidMount(): void {
        this.props.eventCore.on('hover/map', (data) => {
            this.setState({
                currentCoords: {
                    x: data.x,
                    y: data.y
                }
            });
        });
    }

    /******************************/

    render(): JSX.Element {
        return (
            <div className="btn-group mx-1 ifx-as">
                <button
                    className="btn btn-primary ifx-ac"
                >
                    { this.state.currentCoords.x }
                    &nbsp;|&nbsp;
                    { this.state.currentCoords.y }
                </button>
            </div>
        );
    }
}
