// view/components/GoToBtn.tsx
import * as React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import Village from 'core/data/objects/Village';
import { Api } from 'core';

/******************************/

type ComponentProps = Readonly<{
    village: Village,
    map: Api,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{}>

/******************************/

export default
class GoToBtn
extends React.Component<ComponentProps, ComponentState> {
    constructor(props) {
        super(props);

        this.onClickBtn = this.onClickBtn.bind(this);
    }

    /******************************/

    onClickBtn(ev: React.MouseEvent): void {
        let { x, y } = this.props.village;

        this.props.map.events.emit('map/goTo', { x, y });
        ev.stopPropagation();
    }

    /******************************/

    render(): JSX.Element {
        return (
            <button
                className="btn btn-default btn-xs"
                onClick={ this.onClickBtn }
            >
                <FontAwesomeIcon icon="crosshairs" />
            </button>
        );
    }
}
