// view/components/Tooltip.tsx
import * as React from 'react';
import debug from 'debug';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { formatNumberReact } from 'helpers/number';
import { parseSecond } from 'helpers/time';
import { ViewContext } from 'view/viewContext';
import { Hook } from 'libs/ModalManager';
import Item from 'data/data/Item';
import Village from 'data/data/Village';
import Character from 'data/data/Character';
import Tribe from 'data/data/Tribe';

/******************************/

const debugLog = debug('app:view:Tooltip');

/******************************/

type ComponentProps = Readonly<{
    hook: Hook,
    context: ViewContext,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{}>

/******************************/

export default
class Tooltip
extends React.Component<
    ComponentProps,
    ComponentState
> {
    mapMove = false;
    item: Item = null;
    village: Village = null;
    character: Character = null;
    tribe: Tribe = null;

    /******************************/

    onMapMoveChange = (value: boolean): void => {
        let hook = this.props.hook;

        this.mapMove = value;

        if (value) {
            hook.hide();
        } else if (this.item) {
            hook.show();
        }
    }

    onMapEnter = (): void => {
        this.mapMove = false;
        this.props.hook.hide();
    }

    onMapLeave = (): void => {
        this.props.hook.hide();
    }

    onHoverChange = (item: Item): void => {
        let hook = this.props.hook;

        if (item === null) {
            this.item = null;
            hook.hide();
        } else if (item !== this.item) {
            this.item = item;

            if (item instanceof Village) {
                this.village = item;
                this.character = this.village.character;
                this.tribe = this.character && this.character.tribe || null;
            } else if (item instanceof Character) {
                this.village = null;
                this.character = item;
                this.tribe = this.character.tribe;
            } else if (item instanceof Tribe) {
                this.village = null;
                this.character = null;
                this.tribe = item;
            }

            // show
            if (!this.mapMove) {
                hook.show();
                hook.setCursorPosition();
                this.forceUpdate();
            }

            hook.update();
        }
    }

    /******************************/

    componentWillMount(): void {
        let events = this.props.context.mainCanvas.events;

        events.on('changed canvas/move', this.onMapMoveChange);
        events.on('canvas/enter', this.onMapEnter);
        events.on('canvas/leave', this.onMapLeave);
        events.on('changed hover/item', this.onHoverChange);
    }

    componentWillUnmount(): void {
        let events = this.props.context.mainCanvas.events;

        events.removeListener('changed canvas/move', this.onMapMoveChange);
        events.removeListener('canvas/enter', this.onMapEnter);
        events.removeListener('canvas/leave', this.onMapLeave);
        events.removeListener('changed hover/item', this.onHoverChange);
    }

    /******************************/

    render(): JSX.Element {
        debugLog('render');

        return (
            <table className="table table-dark table-responsive table-striped table-sm m-0">
                <tbody>
                    { this.tribe ? (
                        <tr>
                            <td className="icon"><FontAwesomeIcon icon="asterisk" /></td>
                            <td className="text-nowrap">{ this.tribe.name }</td>
                            <td className="value-format">
                                { formatNumberReact(this.tribe.memberCount) }
                                <span className="unit">members</span>
                            </td>
                            <td className="value-format text-right">
                                { formatNumberReact(this.tribe.points) }
                                <span className="unit">p.</span>
                            </td>
                        </tr>
                    ) : null }
                    { this.character ? (
                        <tr>
                            <td className="icon"><FontAwesomeIcon icon="user" /></td>
                            <td className="text-nowrap">{ this.character.name }</td>
                            <td className="value-format">
                                { formatNumberReact(this.character.villageCount) }
                                <span className="unit">villages</span>
                            </td>
                            <td className="value-format text-right">
                                { formatNumberReact(this.character.points) }
                                <span className="unit">p.</span>
                            </td>
                        </tr>
                    ) : null }
                    { this.village ? (
                        <tr>
                            <td className="icon"><FontAwesomeIcon icon="flag" /></td>
                            <td className="text-nowrap">{ this.village.name }</td>
                            <td className="value-format">
                                <span className="special-char">[</span>
                                <span className="value-1">
                                    { this.village.x }
                                </span>
                                <span className="special-char">|</span>
                                <span className="value-1">
                                    { this.village.y }
                                </span>
                                <span className="special-char">]</span>
                            </td>
                            <td className="value-format text-right">
                                { formatNumberReact(this.village.points) }
                                <span className="unit">p.</span>
                            </td>
                        </tr>
                    ) : null }
                </tbody>
            </table>
        );
    }
}
