// view/components/Top20Characters.tsx
import * as React from 'react';
import debug from 'debug';
import { formatNumberReact } from 'helpers/number';
import Character from 'data/data/Character';
import { ViewContext } from 'view/viewContext';
import { IconVillage } from 'view/components/icon';

/******************************/

const debugLog = debug('app:view:top20Characters');

const sortCharacterByPoints = (a: Character, b: Character) => {
    return b.points - a.points;
}

/******************************/

type ComponentProps = Readonly<{
    context: ViewContext
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{}>

/******************************/

export default
class Top20Characters
extends React.Component<
    ComponentProps,
    ComponentState
> {
    update = () => {
        this.forceUpdate();
    }

    componentWillMount(): void {
        this.props.context.data.events.on('changed data/base', this.update);
    }

    componentWillUnmount(): void {
        this.props.context.data.events.removeListener('changed data/base', this.update);
    }

    /******************************/

    onClick(item): void {
        const { context } = this.props;

        context.mainCanvas.events.emit('click/item', item);
    }
    onMouseEnter(item): void {
        const { context } = this.props;

        if (!context.modalSystem.isMoveing()) {
            context.mainCanvas.events.emit('changed hover/item', item);
        }
    }
    onMouseLeave = (e: any): void => {
        const { context } = this.props;

        if (!context.modalSystem.isMoveing()) {
            context.mainCanvas.events.emit('changed hover/item', null);
        }
    }

    /******************************/

    render(): JSX.Element {
        const TOP_SIZE = 25;
        const { context } = this.props;
        const worldDetails = context.data.getWorldDetails();
        const topList = [] as Character[];

        let worst: Character = null;

        function push(character: Character) {
            if (worst === null
                || character.points > worst.points
            ) {
                topList.push(character);
                topList.sort(sortCharacterByPoints);
            }

            if (topList.length > TOP_SIZE) {
                topList.length = TOP_SIZE;
            }

            worst = topList[ topList.length - 1 ] || null;
        }

        for (let character of context.data.getCharacterInterator()) {
            character && push(character);
        }

        let result = [];

        for (let index = 0; index < topList.length; index++) {
            const character = topList[ index ];

            const onClick = () => this.onClick(character);
            const onMouseEnter = () => this.onMouseEnter(character);

            result.push(
                <div
                    key={ `${ index }-position` }
                    className="topTable-position"
                    onClick={ onClick }
                    onMouseEnter={ onMouseEnter }
                    onMouseLeave={ this.onMouseLeave }
                >
                    <span className="unit">#</span>
                    <span className="value-5">{ index + 1 }</span>
                </div>,
                <div
                    key={ `${ index }-name` }
                    className="topTable-name"
                    onClick={ onClick }
                    onMouseEnter={ onMouseEnter }
                    onMouseLeave={ this.onMouseLeave }
                >
                    <span className="value-0">
                        { character.name }
                    </span>
                </div>,
                <div
                    key={ `${ index }-tribe` }
                    className="topTable-points"
                    onClick={ onClick }
                    onMouseEnter={ onMouseEnter }
                    onMouseLeave={ this.onMouseLeave }
                >
                    { character.tribe ? (
                        <>
                            <span className="unit">[</span>
                            <span className="value-4">{ character.tribe.tag }</span>
                            <span className="unit">]</span>
                        </>
                    ) : '-' }
                </div>,
                <div
                    key={ `${ index }-points` }
                    className="topTable-points"
                    onClick={ onClick }
                    onMouseEnter={ onMouseEnter }
                    onMouseLeave={ this.onMouseLeave }
                >
                    { formatNumberReact(character.points) }
                    <span className="unit">p.</span>
                </div>,
                <div
                    key={ `${ index }-off` }
                    className="topTable-off"
                    onClick={ onClick }
                    onMouseEnter={ onMouseEnter }
                    onMouseLeave={ this.onMouseLeave }
                >
                    { formatNumberReact(character.pointsOff) }
                    <span className="value-4">off</span>
                    <span className="unit">p.</span>
                </div>,
                <div
                    key={ `${ index }-def` }
                    className="topTable-def"
                    onClick={ onClick }
                    onMouseEnter={ onMouseEnter }
                    onMouseLeave={ this.onMouseLeave }
                >
                    { formatNumberReact(character.pointsDef) }
                    <span className="value-2">def</span>
                    <span className="unit">p.</span>
                </div>,
                <div
                    key={ `${ index }-bash` }
                    className="topTable-bash"
                    onClick={ onClick }
                    onMouseEnter={ onMouseEnter }
                    onMouseLeave={ this.onMouseLeave }
                >
                    { formatNumberReact(character.pointsTotal) }
                    <span className="value-2">bash</span>
                    <span className="unit">p.</span>
                </div>,
                <div
                    key={ `${ index }-counts` }
                    className="topTable-counts"
                    onClick={ onClick }
                    onMouseEnter={ onMouseEnter }
                    onMouseLeave={ this.onMouseLeave }
                >
                    { formatNumberReact(character.villageCount) }
                    <span className="unit">
                        <IconVillage />
                    </span>
                </div>
            );
        }

        return (
            <div className="topCharacterTable value-format">
                <div>
                    <b className="unit">#</b>
                </div>
                <div>
                    <b className="value-0">Name</b>
                </div>
                <div>
                    <b className="value-0">Tribe</b>
                </div>
                <div>
                    <b className="value-0">Points</b>
                </div>
                <div>
                    <b className="value-0">Off points</b>
                </div>
                <div>
                    <b className="value-0">Off points</b>
                </div>
                <div>
                    <b className="value-0">Bash points</b>
                </div>
                <div>
                    <b className="value-0">Villages</b>
                </div>
                { result }
            </div>
        );
    }
}
