// view/App.tsx
import * as React from 'react';
import {
  BrowserRouter as Router,
  Route,
  RouteComponentProps
} from 'react-router-dom';
import uuid from 'uuid/v4';

// components

import PageMap from 'view/pages/Map';
import PageList from 'view/pages/List';
import createViewContext, { ViewContext } from 'view/viewContext';
import { Context } from 'context';
import GoogleAnalytics from 'view/GoogleAnalytics';
import Gdrp from 'view/Gdrp';
import { getPackage } from 'view/services/package';

/******************************/

type ComponentProps = Readonly<{
    context: Context,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{}>

/******************************/

export default
class App
extends React.Component<ComponentProps, ComponentState> {
    viewContext: ViewContext;

    constructor(props) {
        super(props);
        this.viewContext = createViewContext(this.props.context);
    }

    /******************************/

    render(): JSX.Element {
        let MM = this.viewContext.modalSystem,
            stores = this.viewContext.stores;

        return (
            <>
                <Gdrp context={ this.viewContext } />
                <MM.Container />
                <Router>
                    <Route exact path="/" render={(routeProps) => {
                        setTimeout(() => {
                            stores.state.actions.setMapKey(
                                routeProps.match.params.market,
                                null, null, null
                            );
                        });
                        MM.closeAll();

                        return this.pageList(routeProps);
                    }} />
                    <Route exact path="/:market" render={(routeProps) => {
                        setTimeout(() => {
                            stores.state.actions.setMapKey(
                                routeProps.match.params.market,
                                null, null, null
                            );
                        });
                        MM.closeAll();

                        return this.pageList(routeProps);
                    }} />
                    <Route path="/:market/:worldId/:command?/:value?" render={(routeProps) => {
                        const { market, worldId, command, value } = routeProps.match.params;

                        setTimeout(() => {
                            stores.state.actions.setMapKey(
                                market,
                                worldId,
                                command,
                                value
                            );
                            stores.world.actions.setWorld(
                                market,
                                worldId
                            );

                            if (command === 'package') {
                                getPackage(value)
                                .then(
                                    (newPackage) => {
                                        newPackage.type = 'url';

                                        stores.world.actions.packageAdd(newPackage);
                                        stores.world.actions.packageSelect(newPackage.uuid);
                                    },
                                    () => {
                                        // TODO
                                        alert('invalid data');
                                    }
                                );
                                routeProps.history.replace(`/${ market }/${ worldId }`);
                            }
                        });
                        MM.closeAll();

                        return this.pageMap(routeProps);
                    }} />
                    <GoogleAnalytics context={ this.viewContext } />
                </Router>
            </>
        );
    }

    pageList(
        routeProps: RouteComponentProps
    ): JSX.Element {
        return React.createElement(
            PageList,
            {
                ...routeProps,
                context: this.viewContext
            }
        );
    }

    pageMap(
        routeProps: RouteComponentProps
    ): JSX.Element {
        return React.createElement(
            PageMap,
            {
                ...routeProps,
                context: this.viewContext
            }
        );
    }
}
