// view/App.tsx
import * as React from 'react';
import {
  BrowserRouter as Router,
  Route
} from 'react-router-dom';
import { CanvasApi } from 'canvas';
import createStores from 'stores';
import ModalManager, { Modal, Header, Body, Footer, Hook } from 'libs/ModalManager';

// fortawesome

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { library } from '@fortawesome/fontawesome-svg-core';
import {
    faPowerOff, faBars, faCrosshairs,
    faFlag, faUser, faAsterisk,
    faFire, faClock, faSignal,
    faCaretUp, faCaretDown, faShare, faCog,
    faSearch, faGlobe, faCubes, faHighlighter,
    faExpandArrowsAlt
} from '@fortawesome/free-solid-svg-icons';

// components

import BtnFullSCreen from 'view/components/topBar/BtnFullScreen';
import BtnStats from 'view/components/topBar/BtnStats';
import serviceColorPickerInit from 'view/services/serviceColorPicker';
import featureSettingsInit from 'view/features/modalSettings';
import featureTooltipInit from 'view/features/modalTooltip';
import featureSearchInit from 'view/features/modalSearch';
import featureMenuItem from 'view/features/modalMenuItem';
import { MapDataApi } from 'data';

/******************************/

library.add(
    faPowerOff, faBars, faCrosshairs,
    faFlag, faUser, faAsterisk,
    faFire, faClock, faSignal, faCog,
    faCaretUp, faCaretDown, faShare, faHighlighter,
    faSearch, faGlobe, faCubes, faExpandArrowsAlt
);

/******************************/

type ComponentProps = Readonly<{
    stores: ReturnType<typeof createStores>,
    mainCanvas: CanvasApi,
    appData: MapDataApi
    modalSystem: ModalManager,
    children?: React.ReactNode
}>;
type ComponentState = Readonly<{}>

/******************************/

export default
class App
extends React.Component<ComponentProps, ComponentState> {
    serviceColorPicker: ReturnType<typeof serviceColorPickerInit>;
    featureSettings: ReturnType<typeof featureSettingsInit>;
    featureSearch: ReturnType<typeof featureSearchInit>;

    constructor(props) {
        super(props);
        this.serviceColorPicker = serviceColorPickerInit(this.props.modalSystem);
        this.featureSettings = featureSettingsInit(
            this.props.modalSystem,
            this.props.stores,
            this.props.mainCanvas,
            this.serviceColorPicker
        );
        this.featureSearch = featureSearchInit(
            this.props.modalSystem,
            this.props.mainCanvas,
            this.props.appData
        );

        featureTooltipInit(
            this.props.modalSystem,
            this.props.mainCanvas
        );
        featureMenuItem(
            this.props.modalSystem,
            this.props.mainCanvas
        );
    }

    /******************************/

    componentDidMount(): void {
        let canvasContainer: HTMLElement = this.refs.canvasContainer as HTMLElement;

        canvasContainer.appendChild(this.props.mainCanvas.canvas);
    }

    /******************************/

    render(): JSX.Element {
        let MM = this.props.modalSystem;

        return (
            <Router>
                {/* <div>
                    <Route exact path="/" render={(props) => (
                        'Home'
                    )} />
                    <Route exact path="/:market" render={(props) => (
                        <pre>
                            { JSON.stringify(props, null, '') }
                        </pre>
                    )} />
                    <Route path="/:market/:worldId" render={(props) => (
                        <pre>
                            { JSON.stringify(props, null, '') }
                        </pre>
                    )} />
                </div> */}
                <MM.Container />
                <div className="layout-map">
                    <div className="top-bar">
                        <div className="left">
                            <button
                                className="btn btn-primary ifx-ac mx-1"
                            >
                                <FontAwesomeIcon icon="bars" />
                            </button>
                            <button
                                className="btn btn-primary ifx-ac mx-1"
                            >
                                <FontAwesomeIcon icon="highlighter" />
                            </button>
                            <button
                                className="btn btn-primary ifx-ac mx-1"
                                onClick={ () => this.featureSearch() }
                            >
                                <FontAwesomeIcon icon="search" />
                            </button>
                        </div>
                        <div className="right">
                            <BtnStats
                                mainCanvas={ this.props.mainCanvas }
                                worldStore={ this.props.stores.world }
                            />
                            <BtnFullSCreen />

                            <button
                                className="btn btn-primary ifx-ac mx-1"
                                onClick={ (e) => this.featureSettings() }
                            >
                                <FontAwesomeIcon icon="cog" />
                            </button>

                            <button
                                className="btn btn-primary ifx-ac mx-1"
                            >
                                <FontAwesomeIcon icon="share" />
                            </button>
                        </div>
                    </div>
                    <div className="canvas">
                        <div ref="canvasContainer"></div>
                    </div>
                </div>
            </Router>
        );
    }
}
