// common/MapColors.ts
export const COLOR_INDEX_BACKGROUND = 0;
export const COLOR_INDEX_BORDER_PROVINCE = 1;
export const COLOR_INDEX_BORDER_CONTINENT = 2;
export const COLOR_INDEX_VILLAGE_PLAYER = 3;
export const COLOR_INDEX_VILLAGE_BARBARIAN = 4;
export const COLOR_INDEX_HOVER_ITEM = 5;
export const COLOR_INDEX_HOVER_GROUP = 6;

type MapColorItem = {
    index: number,
    label: string,
    key: string
};

export const mapColorList: MapColorItem[] = [
    {
        index: COLOR_INDEX_BACKGROUND,
        label: 'Background',
        key: 'colorBackground'
    },
    {
        index: COLOR_INDEX_BORDER_PROVINCE,
        label: 'Province border',
        key: 'colorBorderProvince'
    },
    {
        index: COLOR_INDEX_BORDER_CONTINENT,
        label: 'Continent border',
        key: 'colorBorderContinent'
    },
    {
        index: COLOR_INDEX_VILLAGE_PLAYER,
        label: 'Player\'s villages',
        key: 'colorVillagePlayer'
    },
    {
        index: COLOR_INDEX_VILLAGE_BARBARIAN,
        label: 'Barbarian\'s villages',
        key: 'colorVillageBarbarian'
    },
    {
        index: COLOR_INDEX_HOVER_ITEM,
        label: 'Hover of item',
        key: 'colorHoverItem'
    },
    {
        index: COLOR_INDEX_HOVER_GROUP,
        label: 'Hover of group',
        key: 'colorHoverGroup'
    }
];

const map_index2color: {[key: number]: MapColorItem} = {};
for (let item of mapColorList) {
    map_index2color[ item.index ] = item;
}

const map_key2color: {[key: string]: MapColorItem} = {};
for (let item of mapColorList) {
    map_key2color[ item.key ] = item;
}

export const maps = {
    index2color: map_index2color,
    key2color: map_key2color
}

/******************************/

export default class MapColors {
    colorBackground = '#004000';
    colorBorderProvince = '#888888';
    colorBorderContinent = '#ffff00';
    colorVillagePlayer = '#003000';
    colorVillageBarbarian = '#005000';
    colorHoverItem = '#ffffff';
    colorHoverGroup = '#999999';
}
