// stores/world/@/markers/item.ts
import uuid from 'uuid/v4';
import * as $marker from "stores/world/@/markers/marker";
import * as $package from "stores/world/@/markers/package";

/******************************/

export function createItem(
    name: string,
    type: string,
    itemId: number,
    parent?: string
) {
    return {
        uuid: uuid(),
        parent: parent || null,
        name, type, itemId,
        active: true,
    };
}

/******************************/

export function wrapper(
    itemObject: Item,
    packageObject: $package.WPackage
): WItem {
    Object.setPrototypeOf(itemObject, ItemPrototype);

    Object.hasOwnProperty.call(itemObject, 'refPackage')
        ? itemObject['refPackage'] = packageObject
        : Object.defineProperty(itemObject, 'refPackage', {value: packageObject, writable: true})

    return itemObject as WItem;
}

const ItemPrototype = {
    clone() {
        return wrapper({ ...this }, this.refPackage);
    },
    getParent() {
        const refPackage = this.refPackage as $package.WPackage;

        try {
            return refPackage.getMarkerByUuid(this.parent);
        } catch (error) {
            const rootMarker = refPackage.getRootMarker();

            this.parent = rootMarker.uuid;

            return rootMarker;
        }
    },
    getColor(
        colorIndex: 1 | 2 | 3 = 1
    ): {
        value: string,
        index: string,
        uuid: string
    } {
        const parent = this.getParent() as $marker.WMarker;

        if (parent.parent === null) {
            return parent[`color${ colorIndex }`].active
                ? {
                    value: parent[`color${ colorIndex }`].value,
                    index: parent.index,
                    uuid: parent.uuid
                }
                : null;
        }

        return parent.getColor(colorIndex);
    },
    isActive(): boolean {
        return this.active
            && this.getParent().isActive();
    },
    isActiveSelf(): boolean {
        return this.active;
    }
};

export type Item = ReturnType<typeof createItem>;
export type WItem = Item & typeof ItemPrototype;
